cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1269,6 +1353,55 @@ void reader::impl::preprocess_pages(size_t skip_rows,
   auto& chunks = _file_itm_data.chunks;
   auto& pages  = _file_itm_data.pages_info;
 
+  // compute page ordering.
+  //
+  // ordering of pages is by input column schema, repeated across row groups.  so
+  // if we had 3 columns, each with 2 pages, and 1 row group, our schema values might look like
+  //
+  // 1, 1, 2, 2, 3, 3
+  //
+  // However, if we had more than one row group, the pattern would be
+  //
+  // 1, 1, 2, 2, 3, 3, 1, 1, 2, 2, 3, 3
+  // ^ row group 0     |
+  //                   ^ row group 1
+  //
+  // To process pages by key (exclusive_scan_by_key, reduce_by_key, etc), the ordering we actually
+  // want is
+  //
+  // 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3
+  //
+  // We also need to preserve key-relative page ordering, so we need to use a stable sort.
+  rmm::device_uvector<int> page_keys(pages.size(), _stream);
+  rmm::device_uvector<int> page_index(pages.size(), _stream);
+  {
+    thrust::transform(rmm::exec_policy(_stream),
+                      pages.device_ptr(),
+                      pages.device_ptr() + pages.size(),
+                      page_keys.begin(),
+                      get_page_schema{});
+
+    thrust::sequence(rmm::exec_policy(_stream), page_index.begin(), page_index.end());
+    thrust::stable_sort_by_key(rmm::exec_policy(_stream),
+                               page_keys.begin(),
+                               page_keys.end(),
+                               page_index.begin(),
+                               thrust::less<int>());
+  }
+
+  // detect malformed columns.
+  // - we have seen some cases in the wild where we have a row group containing N
+  //   rows, but the total number of rows in the pages for column X is != N. while it
+  //   is possible to load this by just capping the number of rows read, we cannot tell
+  //   which rows are invalid so we may be returning bad data. in addition, this mismatch
+  //   confuses the chunked reader
+  detect_malformed_pages(pages,
+                         chunks,
+                         page_keys,
+                         page_index,
+                         uses_custom_row_bounds ? std::nullopt : std::make_optional(num_rows),
+                         _stream);
```

Comment:
This seems expensive to me instead of "lightweight". Do you know at what scale this computation will be, comparing to the number of rows?

Reply:
Ah, is this just moved up from below?

Reply:
Correct.  The sorting of pages by schema is needed for everything.